src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ClassDataContract.cs

```diff
@@ -1482,70 +1288,210 @@ internal struct Member
+                                return false;
+                        }
+                        else
+                        {
+                            Dictionary<string, DataMember> membersDictionary = new Dictionary<string, DataMember>(Members.Count);
+                            List<DataMember> dataContractMembersList = new List<DataMember>();
+                            for (int i = 0; i < Members.Count; i++)
+                            {
+                                membersDictionary.Add(Members[i].Name, Members[i]);
+                            }
+
+                            for (int i = 0; i < dataContract.Members.Count; i++)
+                            {
+                                // check that all datamembers common to both datacontracts match
+                                if (membersDictionary.TryGetValue(dataContract.Members[i].Name, out DataMember? dataMember))
+                                {
+                                    if (dataMember.Equals(dataContract.Members[i], checkedContracts))
+                                    {
+                                        membersDictionary.Remove(dataMember.Name);
+                                    }
+                                    else
+                                    {
+                                        return false;
+                                    }
+                                }
+                                // otherwise save the non-matching datamembers for later verification
+                                else
+                                {
+                                    dataContractMembersList.Add(dataContract.Members[i]);
+                                }
+                            }
+
+                            // check that datamembers left over from either datacontract are optional
+                            if (!IsEveryDataMemberOptional(membersDictionary.Values))
+                                return false;
+                            if (!IsEveryDataMemberOptional(dataContractMembersList))
+                                return false;
+                        }
+                    }
+
+                    if (BaseClassContract == null)
+                        return (dataContract.BaseClassContract == null);
+                    else if (dataContract.BaseClassContract == null)
+                        return false;
+                    else
+                        return BaseClassContract.Equals(dataContract.BaseClassContract, checkedContracts);
+                }
+            }
+            return false;
+        }
+
+        private static bool IsEveryDataMemberOptional(IEnumerable<DataMember> dataMembers)
+        {
+            foreach (DataMember dataMember in dataMembers)
+            {
+                if (dataMember.IsRequired)
+                    return false;
+            }
+            return true;
+        }
+
+        public override int GetHashCode()
+        {
```

Comment:
Might be worth adding a comment explaining that the compiler throws a small tantrum if you override Equals and not GetHashCode